library/std/src/sys/unix/fs.rs

```diff
@@ -165,31 +166,44 @@ cfg_has_statx! {{
             ) -> c_int
         }
 
-        match STATX_STATE.load(Ordering::Relaxed) {
-            0 => {
-                // It is a trick to call `statx` with null pointers to check if the syscall
-                // is available. According to the manual, it is expected to fail with EFAULT.
-                // We do this mainly for performance, since it is nearly hundreds times
-                // faster than a normal successful call.
-                let err = cvt(statx(0, ptr::null(), 0, libc::STATX_ALL, ptr::null_mut()))
-                    .err()
-                    .and_then(|e| e.raw_os_error());
-                // We don't check `err == Some(libc::ENOSYS)` because the syscall may be limited
-                // and returns `EPERM`. Listing all possible errors seems not a good idea.
-                // See: https://github.com/rust-lang/rust/issues/65662
-                if err != Some(libc::EFAULT) {
-                    STATX_STATE.store(1, Ordering::Relaxed);
-                    return None;
-                }
-                STATX_STATE.store(2, Ordering::Relaxed);
-            }
-            1 => return None,
-            _ => {}
+        if STATX_SAVED_STATE.load(Ordering::Relaxed) == STATX_STATE::Unavailable as u8 {
+            return None;
         }
 
         let mut buf: libc::statx = mem::zeroed();
         if let Err(err) = cvt(statx(fd, path, flags, mask, &mut buf)) {
-            return Some(Err(err));
+            if STATX_SAVED_STATE.load(Ordering::Relaxed) == STATX_STATE::Present as u8 {
+                return Some(Err(err));
+            }
```

Comment:
It is not quite correct to rely on this because syscall filters can be installed during program runtime. So the behavior can change. At that point ENOSYS could bubble up to user code because the flag now says all errors should be bubbled up.

If ENOSYS and EPERM can be reliably distinguished from other errors then updating the atomic on any error is better, as we do for [splice and sendfile](https://github.com/rust-lang/rust/blob/754f6d4a8cf1963a3446b20b20fd10f42df1ee41/library/std/src/sys/unix/kernel_copy.rs#L666-L672)
If EPERM can also convey a real error then the pre-probing behavior that existed before this PR is correct.

Reply:
I don't think there is a real issue here.

Detection only showed up because *some* deployments got rust utilizing statx running on kernels which did not have it *yet* or otherwise did not have configuration updated to cope with it, resulting in EPERM from seccomp.

To my understanding statx is the way to stat going forward and as such disabling it *on purpose* is a misconfiguration of the system, on par with disabling a syscall like open.

Reply:
Any sane seccomp filters are deny-by-default, which means any filter list that doesn't explicitly opt into `statx` can end up blocking the call.
Combine that with the fact that people run outdated software (which can contain equally outdated filter lists) it is possible that you end up with a situation where stat is allowed but statx isn't.

And what happens in the wild is not really relevant for correctness. For correctness it only matters that syscall availability can change at runtime.

[People do all kinds of crazy stuff](https://github.com/rust-lang/rust/issues/96621) and we need to be defensive.

Reply:
Just in case I have to stress that with and without the patch the rust binary will ultimately detect statx (un)availabilty the same way, whether the kernel does not support statx to begin with or seccomp is configured to disable it, as long as the latter happened prior to running detection code.

I have to concede it is possible someone has outdated seccomp filters, which happen to be disabled when a rust binary starts execing and only get enabled later. In such a case errors from seccomp will indeed be returned. But this was already the case with the code prior to me patching it (read: no regression on that front).

I also feel inclined to note the sendfile example does not "update the atomic every time". The code speculatively executes the syscall, presumably to avoid paying specifically for detection, and the atomic store is a one time ordeal per exec.

If anything my patch moved statx handling closer to that spirit, but in contrast it still does not handle such a failure at an arbitrary moment.

Now, legally encountered permission problems return EACCESS. The kernel has some hairy code to support all of this and I definitely would not stack my neck on EPERM being only returned by seccomp.

All that said, as mentioned in another comment, I plan to add post a followup to fix the pre-existing bugs I mentioned in FIXME comments. If it does not get hairy, I may as well throw an extra for seccomp.

Reply:
Yes, totally agreed, it's an edge-case. But seccomp filters getting installed during the runtime of a program after doing some setup work is definitely a thing that happens in daemons that lock themselves down.

> All that said, as mentioned in another comment, I plan to post a followup to fix the pre-existing bugs

👍